cmd/kubectl-gadget/utils/trace.go

```diff
@@ -463,95 +466,157 @@ func getTraceWatcher(traceID string) (watch.Interface, error) {
+
+				trace, ok := event.Object.(*gadgetv1alpha1.Trace)
+				if !ok {
+					return false, fmt.Errorf("Received object cannot be cast to Trace: %v", event.Object)
+				}
 
-			nodeErrors[trace.Spec.Node] = trace.Status.OperationError
+				traceName := trace.ObjectMeta.Name
+
+				// We also remove it from the maps to avoid returning a deleted trace
+				// and timeing out.
+				delete(satisfiedTraces, traceName)
+				delete(erroredTraces, traceName)
+
+				return false, nil
+			case watch.Modified:
+				// We will deal with this type of event below
+			case watch.Error:
+				// Deal particularly with error.
+				return false, fmt.Errorf("Received event is an error one: %v", event)
+			case watch.Added:
+				// createTraces() creates traces synchronously.
+				// So, if a watch.Added event occurs it means there is a problem (e.g.
+				// the user creates a trace by snooping on the traceID of existing
+				// traces).
+				return false, fmt.Errorf("No traces with the given traceID (%s) should be created", traceID)
+			default:
+				// We are not interested in other event types.
+				return false, nil
+			}
 
-			return false, nil
-		}
+			trace, ok := event.Object.(*gadgetv1alpha1.Trace)
+			if !ok {
+				return false, fmt.Errorf("Received object cannot be cast to Trace: %v", event.Object)
+			}
 
-		// If the trace does not satisfy the condition function, we are not
-		// interested.
-		if !conditionFunction(trace) {
-			return false, nil
-		}
+			if trace.Status.OperationWarning != "" {
+				// The trace can have a warning but satisfies conditionFunction.
+				// So, we do not add it to the map here.
+				nodeWarnings[trace.Spec.Node] = trace.Status.OperationWarning
+			}
 
-		watchedTracesNumber++
-		// If the current trace matches our filter we add it to the list of trace
-		// we will return.
-		returnedTraces.Items = append(returnedTraces.Items, *trace)
+			if trace.Status.OperationError != "" {
+				erroredTraces[trace.ObjectMeta.Name] = trace
 
-		return watchedTracesNumber == tracesNumber, nil
-	})
-	cancel()
+				// If the trace satisfied the function, we do not care now because it
+				// has an error.
+				delete(satisfiedTraces, trace.ObjectMeta.Name)
+
+				return false, nil
+			}
+
+			// If the trace does not satisfy the condition function, we are not
+			// interested.
+			if !conditionFunction(trace) {
+				return false, nil
+			}
+
+			satisfiedTraces[trace.ObjectMeta.Name] = trace
+
+			return len(satisfiedTraces)+len(erroredTraces) == tracesToWatch, nil
+		})
+		cancel()
+	}
+
+	for _, trace := range erroredTraces {
+		nodeErrors[trace.Spec.Node] = trace.Status.OperationError
+	}
 
 	// We print errors whatever happened.
-	defer printTraceFeedback(nodeErrors, tracesNumber)
+	printTraceFeedback(nodeErrors, tracesNumber)
 
 	// We print warnings only if all trace failed.
-	if len(returnedTraces.Items) == 0 {
+	if len(satisfiedTraces) == 0 {
 		printTraceFeedback(nodeWarnings, tracesNumber)
 	}
 
 	if err != nil {
```

Comment:
`err` is not the good one.

Reply:
Why not?

Reply:
I wrote this comment to remind me to check this behavior.
But in the `watch`, `err` value is just changed and it nos declared again.
So, it is fine to use it here.

Reply:
`err` is wrongly being redeclared [here](https://github.com/kinvolk/inspektor-gadget/blob/e6ff7773f65674691ff850a02401f01f9633effb/cmd/kubectl-gadget/utils/trace.go#L514) so this error check is useless right now. I think that by reusing the same outer `err` variable, the logic is correct.

Reply:
> err is wrongly being redeclared [here](https://github.com/kinvolk/inspektor-gadget/blob/e6ff7773f65674691ff850a02401f01f9633effb/cmd/kubectl-gadget/utils/trace.go#L514) so this error check is useless right now. I think that by reusing the same [outer err variable](https://github.com/kinvolk/inspektor-gadget/blob/e6ff7773f65674691ff850a02401f01f9633effb/cmd/kubectl-gadget/utils/trace.go#L475), the logic is correct.

I do not think it is redeclared, `watcher` is declared here but `err` should be the one already declared above.
So, I think the actual logic is OK.
Otherwise, I will make it clearer defining `extError` and `innerError`.

Reply:
> I do not think it is redeclared, `watcher` is declared here but `err` should be the one already declared above.
>
As far as I know, if both [[1](https://github.com/kinvolk/inspektor-gadget/blob/032015089c5001c2ecaa58040a353421983bf54b/cmd/kubectl-gadget/utils/trace.go#L475)] and [[2]](https://github.com/kinvolk/inspektor-gadget/blob/032015089c5001c2ecaa58040a353421983bf54b/cmd/kubectl-gadget/utils/trace.go#L514)  were in the same block, then what you are saying would be true. Example [here](https://go.dev/play/p/HV_sDKtSK6O). Ref: https://go.dev/ref/spec#Short_variable_declarations.

However, given that they are within different blocks, it means that inner `err` is a new variable and not a redeclaration. Example [here](https://go.dev/play/p/NCSLdgITUXr). 

Or what did I miss/misunderstand?

Reply:
> Or what did I miss/misunderstand?

Nothing, this pattern is just particularly unclear and I will then make it clearer by adding two differently named variables.

Reply:
IMHO, having two error variables makes things even more complex 😅 .

I think it is enough to first declare the watcher and user `=` instead of `:=` when calling `getTraceWatcher()`. In that way, you are sure it is using the same outer `err` variable: 

```
		var watcher watch.Interface
		watcher, err = getTraceWatcher(traceID, traceList.ListMeta.ResourceVersion)		
```

Reply:
I agree that having two error variables is even more complicated.